src/backend/distributed/utils/background_jobs.c

```diff
@@ -310,6 +335,422 @@ CitusBackgroundTaskQueueMonitorErrorCallback(void *arg)
+								  "Already reached the maximum number of task "
+								  "executors: %ld/%d",
+								  queueMonitorExecutionContext->currentExecutorCount,
+								  MaxBackgroundTaskExecutors)));
+			queueMonitorExecutionContext->backgroundWorkerFailedStartTime =
+				GetCurrentTimestamp();
+		}
+
+		const long delay_ms = 1000;
+		(void) WaitLatch(MyLatch, WL_LATCH_SET | WL_TIMEOUT |
+						 WL_EXIT_ON_PM_DEATH,
+						 delay_ms, WAIT_EVENT_PG_SLEEP);
+		ResetLatch(MyLatch);
+
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * NewExecutorExceedsPgMaxWorkers returns true if currently we reached Postgres' max worker count.
+ * It also sleeps 1 sec to let running tasks progress so that we have better chance to not hit
+ * that limit again.
+ */
+static bool
+NewExecutorExceedsPgMaxWorkers(BackgroundWorkerHandle *handle,
+							   QueueMonitorExecutionContext *queueMonitorExecutionContext)
+{
+	if (handle == NULL)
+	{
+		/*
+		 * we are unable to start a background worker for the task execution.
+		 * Probably we are out of background workers. Warn for the first failure
+		 * after a successful worker allocation happened, that is, we do not warn if
+		 * we repeatedly come here without a successful worker allocation.
+		 */
+		if (queueMonitorExecutionContext->backgroundWorkerFailedStartTime == 0)
+		{
+			ereport(WARNING, (errmsg(
+								  "unable to start background worker for "
+								  "background task execution"),
+							  errdetail(
+								  "Current number of task "
+								  "executors: %ld/%d",
+								  queueMonitorExecutionContext->currentExecutorCount,
+								  MaxBackgroundTaskExecutors)));
+			queueMonitorExecutionContext->backgroundWorkerFailedStartTime =
+				GetCurrentTimestamp();
+		}
+
+		const long delay_ms = 1000;
+		(void) WaitLatch(MyLatch, WL_LATCH_SET | WL_TIMEOUT |
+						 WL_EXIT_ON_PM_DEATH,
+						 delay_ms, WAIT_EVENT_PG_SLEEP);
+		ResetLatch(MyLatch);
+
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * AssignRunnableTaskToNewExecutor tries to assign given runnable task to a new task executor.
+ * It reports the assignment status as return value.
+ */
+static bool
+AssignRunnableTaskToNewExecutor(BackgroundTask *runnableTask,
+								QueueMonitorExecutionContext *queueMonitorExecutionContext)
+{
+	Assert(runnableTask && runnableTask->status == BACKGROUND_TASK_STATUS_RUNNABLE);
+
+	if (NewExecutorExceedsCitusLimit(queueMonitorExecutionContext))
+	{
+		/* escape if we hit citus executor limit */
+		return false;
+	}
+
+	char *databaseName = get_database_name(MyDatabaseId);
+	char *userName = GetUserNameFromId(runnableTask->owner, false);
+
+	/* try to create new executor and make it alive during queue monitor lifetime */
+	MemoryContext oldContext = MemoryContextSwitchTo(queueMonitorExecutionContext->ctx);
+	dsm_segment *seg = NULL;
+	BackgroundWorkerHandle *handle = StartCitusBackgroundTaskExecutor(databaseName,
+																	  userName,
+																	  runnableTask->
+																	  command,
```

Comment:
nit: breaking pointer dereferences like this across line isn't great to read. Sadly our autoformatting does that when there's not enough space on the line. If you put a newline between `StartCitusBackgroundTaskExecutor(` and `databaseName` it should be able to format it better.